src/network/networkd-dhcp-server.c

```diff
@@ -570,3 +579,101 @@ int config_parse_dhcp_server_lpr_servers(
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        _cleanup_(sd_dhcp_static_lease_unrefp) sd_dhcp_static_lease *lease = NULL;
+        _cleanup_free_ char *word = NULL;
+        _cleanup_free_ struct ether_addr *n = NULL;
+        OrderedHashmap **static_leases = data;
+        union in_addr_union addr;
+        const char *p;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        if (isempty(rvalue)) {
+                *static_leases = ordered_hashmap_free(*static_leases);
+                return 0;
+        }
+
+        p = rvalue;
+        r = extract_first_word(&p, &word, " ", 0);
+        if (r == -ENOMEM)
+                return log_oom();
+        if (r <= 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Invalid hardware address, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+
+        n = new(struct ether_addr, 1);
+        if (!n)
+                return log_oom();
+
+        r = ether_addr_from_string(word, n);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Not a valid MAC address, ignoring: %s", word);
+                return 0;
+        }
+
+        word = mfree(word);
+        r = extract_first_word(&p, &word, " ", 0);
+        if (r == -ENOMEM)
+                return log_oom();
+        if (r <= 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Invalid IP address, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+
+        r = in_addr_from_string(AF_INET, word, &addr);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Failed to parse DHCPv4 IPv4 address data, ignoring assignment: %s", word);
+                return 0;
+        }
+
+        r = sd_dhcp_static_lease_new_by_mac(n->ether_addr_octet, ETH_ALEN, &addr.in, &lease);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Failed to store DHCPv4 static lease '%s', ignoring assignment: %m", rvalue);
+                return 0;
+        }
+
+        r = ordered_hashmap_ensure_allocated(static_leases, &dhcp_static_leases_hash_ops);
+        if (r < 0)
+                return log_oom();
+
+        /* Overwrite existing option */
+        r = ordered_hashmap_replace(*static_leases, &lease->client_id, lease);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Failed to store DHCPv4 static lease '%s', ignoring assignment: %m", rvalue);
+                return 0;
+        }
+
+        sd_dhcp_static_lease *removed_lease = NULL;
+        removed_lease = ordered_hashmap_remove(*static_leases, &lease->client_id);
```

Comment:
i don't get it? you insert the entry above, and remove it right-away again here? huh?

Reply:
I had some confusion with ordered_hashmap_remove_and_replace and ordered_hashmap_remove_and_put, as they return an error if the entry in the hashmap does not exist. So i fixed it now.